### ISSUE TITLE
Selenium 3.0.1+ Firefox 50.0 test would have connection refused

### ISSUE THREAD

**Issue by HackToday** [mozilla/geckodriver#432]
If you are having an issue with GeckoDriver starting up, please make sure that your client bindings support using it.

## Meta -
OS:  CentOS 7.2
e.g.: Windows 10? OSX?

Browser Version:   Firefox 50.0
e.g.: 50.0.2623.87 (64-bit)

## Expected Behavior -
Test would run OK

## Actual Behavior -
Sometimes test would fail with connection refused, it is clearly not test issue, it is possible about selenium or Firefox issue

## Steps to reproduce -
_Note:_ Please be sure to include an SSCCE (Short, Self Contained, Correct [compilable] example) http://sscce.org/
If you can't provide a link to the page, consider creating a reproducible page on https://jsfiddle.net/

Can not give a easy steps to reproduce, it can sometime happen. (maybe 3 in 10 times)

```
request = <SubRequest 'driver' for <Function 'test_app_rs_edit_port'>>
driver_class = <class 'selenium.webdriver.firefox.webdriver.WebDriver'>
driver_kwargs = {'firefox_profile': <selenium.webdriver.firefox.firefox_profile.FirefoxProfile object at 0x2603c10>}

    @pytest.yield_fixture
    def driver(request, driver_class, driver_kwargs):
        """Returns a WebDriver instance based on options and capabilities"""
>       driver = driver_class(**driver_kwargs)

/usr/lib/python2.7/site-packages/pytest_selenium/pytest_selenium.py:107: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python2.7/site-packages/selenium/webdriver/firefox/webdriver.py:145: in __init__
    keep_alive=True)
/usr/lib/python2.7/site-packages/selenium/webdriver/remote/webdriver.py:92: in __init__
    self.start_session(desired_capabilities, browser_profile)
/usr/lib/python2.7/site-packages/selenium/webdriver/remote/webdriver.py:179: in start_session
    response = self.execute(Command.NEW_SESSION, capabilities)
/usr/lib/python2.7/site-packages/selenium/webdriver/remote/webdriver.py:236: in execute
    self.error_handler.check_response(response)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <selenium.webdriver.remote.errorhandler.ErrorHandler object at 0x273fad0>
response = {'status': 500, 'value': '{"error":"unknown error","message":"connection refused"}'}

    def check_response(self, response):
        """
            Checks that a JSON response from the WebDriver does not have an error.
    
            :Args:
             - response - The JSON response from the WebDriver server as a dictionary
               object.
    
            :Raises: If the response contains an error message.
            """
        status = response.get('status', None)
        if status is None or status == ErrorCode.SUCCESS:
            return
    
        value = None
        message = response.get("message", "")
        screen = response.get("screen", "")
        stacktrace = None
        if isinstance(status, int):
            value_json = response.get('value', None)
            if value_json and isinstance(value_json, basestring):
                import json
                try:
                    value = json.loads(value_json)
                    status = value.get('error', None)
                    if status is None:
                        status = value["status"]
                        message = value["value"]
                        if not isinstance(message, basestring):
                            value = message
                            try:
                                message = message['message']
                            except TypeError:
                                message = None
                    else:
                        message = value.get('message', None)
                except ValueError:
                    pass
    
        exception_class = ErrorInResponseException
        if status in ErrorCode.NO_SUCH_ELEMENT:
            exception_class = NoSuchElementException
        elif status in ErrorCode.NO_SUCH_FRAME:
            exception_class = NoSuchFrameException
        elif status in ErrorCode.NO_SUCH_WINDOW:
            exception_class = NoSuchWindowException
        elif status in ErrorCode.STALE_ELEMENT_REFERENCE:
            exception_class = StaleElementReferenceException
        elif status in ErrorCode.ELEMENT_NOT_VISIBLE:
            exception_class = ElementNotVisibleException
        elif status in ErrorCode.INVALID_ELEMENT_STATE:
            exception_class = InvalidElementStateException
        elif status in ErrorCode.INVALID_SELECTOR \
                or status in ErrorCode.INVALID_XPATH_SELECTOR \
                or status in ErrorCode.INVALID_XPATH_SELECTOR_RETURN_TYPER:
            exception_class = InvalidSelectorException
        elif status in ErrorCode.ELEMENT_IS_NOT_SELECTABLE:
            exception_class = ElementNotSelectableException
        elif status in ErrorCode.INVALID_COOKIE_DOMAIN:
            exception_class = WebDriverException
        elif status in ErrorCode.UNABLE_TO_SET_COOKIE:
            exception_class = WebDriverException
        elif status in ErrorCode.TIMEOUT:
            exception_class = TimeoutException
        elif status in ErrorCode.SCRIPT_TIMEOUT:
            exception_class = TimeoutException
        elif status in ErrorCode.UNKNOWN_ERROR:
            exception_class = WebDriverException
        elif status in ErrorCode.UNEXPECTED_ALERT_OPEN:
            exception_class = UnexpectedAlertPresentException
        elif status in ErrorCode.NO_ALERT_OPEN:
            exception_class = NoAlertPresentException
        elif status in ErrorCode.IME_NOT_AVAILABLE:
            exception_class = ImeNotAvailableException
        elif status in ErrorCode.IME_ENGINE_ACTIVATION_FAILED:
            exception_class = ImeActivationFailedException
        elif status in ErrorCode.MOVE_TARGET_OUT_OF_BOUNDS:
            exception_class = MoveTargetOutOfBoundsException
        else:
            exception_class = WebDriverException
        if value == '' or value is None:
            value = response['value']
        if isinstance(value, basestring):
            if exception_class == ErrorInResponseException:
                raise exception_class(response, value)
            raise exception_class(value)
        if message == "" and 'message' in value:
            message = value['message']
    
        screen = None
        if 'screen' in value:
            screen = value['screen']
    
        stacktrace = None
        if 'stackTrace' in value and value['stackTrace']:
            stacktrace = []
            try:
                for frame in value['stackTrace']:
                    line = self._value_or_default(frame, 'lineNumber', '')
                    file = self._value_or_default(frame, 'fileName', '<anonymous>')
                    if line:
                        file = "%s:%s" % (file, line)
                    meth = self._value_or_default(frame, 'methodName', '<anonymous>')
                    if 'className' in frame:
                        meth = "%s.%s" % (frame['className'], meth)
                    msg = "    at %s (%s)"
                    msg = msg % (meth, file)
                    stacktrace.append(msg)
            except TypeError:
                pass
        if exception_class == ErrorInResponseException:
            raise exception_class(response, message)
        elif exception_class == UnexpectedAlertPresentException and 'alert' in value:
            raise exception_class(message, screen, stacktrace, value['alert'].get('text'))
>       raise exception_class(message, screen, stacktrace)
E       WebDriverException: Message: connection refused

/usr/lib/python2.7/site-packages/selenium/webdriver/remote/errorhandler.py:192: WebDriverException

```

As issue here https://github.com/SeleniumHQ/selenium/issues/3201  selenium guys said `The connection refused would be happening when trying to communicate with GeckoDriver`

 - [x] A [trace level log](https://github.com/mozilla/geckodriver#firefox-capabilities) from geckodriver showing the issue
 - [ ] A minimal HTML file (or files) that can be used to reproduce the issue
 - [ ] A Selenium code snippet that can be run as is to reproduce the issues

** If any of the above are missing we will have to unforunately close your issue. We will gladly reopen the issue once all the information requested has been added **

**Comment by HackToday** [comment]
The original issue was  https://github.com/mozilla/geckodriver/issues/417
As @andreastt  required,  trace log is needed for debug, I have reproduced with trace level logs, and details can be found here (as trace log is very large)

https://gist.github.com/HackToday/b39d3ea7b20a90b327b9e011cf8bb3d1

And netstat and ps both show no two geckodriver running.

**Comment by HackToday** [comment]
And cc  @jgraham  if can help debug this issue.  Thanks

**Comment by cgoldberg** [comment]
duplicate of #417. 

@HackToday , If you have additional information, please provide it in the original issue so it can be re-opened.

**Comment by cgoldberg** [comment]
@HackToday 
Your stacktrace clearly shows that it is failing in selenium webdriver code (when creating a new webdriver session).  Selenium WebDriver bugs/issues do not belong in the GeckoDriver issue tracker.  However, I'm not even sure this is a valid selenium issue.  It looks like you just wrote some bad tests or you have issues in your environment under test.  Did you read the stacktrace you posted?

Specifically, your error shows that webdriver's `UnexpectedAlertPresentException` is thrown and not handled.  That means a modal alert with the message "connection refused" is blocking your UI and you haven't instructed webdriver how to handle this.

You must either:
 * handle the exception when your test gets blocked by this alert.
or 
 * figure out the root cause of why this unexpected alert is being presented, and suppress it.

either way, I don't see how geckodriver or selenium project maintainers could possibly fix this for you...  especially without any evidence that a problem even exists on their end.

**Comment by HackToday** [comment]
Hi @cgoldberg  Thanks for your reply. Let me reply your questions and comments:

1> I opened new issue, because when I tried to reply in closed issue, but seems nobody cared the new log or trace info.

2> Also the root cause for above selenium test issue, it seems not test itself. As test follows selenium test, just open browser and click some buttons and close browser and etc.

The geckodriver logs (https://gist.githubusercontent.com/HackToday/b39d3ea7b20a90b327b9e011cf8bb3d1/raw/4c30cce5702ce7f78f0847d433ed907cc2daabb7/geckodrivertrace-log ) show that 

```
1480494227580	addons.manager	DEBUG	Provider finished startup: GMPProvider
1480494227580	addons.manager	DEBUG	Starting provider: PluginProvider
1480494227580	addons.manager	DEBUG	Registering shutdown blocker for PluginProvider
1480494227581	addons.manager	DEBUG	Provider finished startup: PluginProvider
1480494227581	addons.manager	DEBUG	Completed startup sequence
1480494227805	Marionette	ERROR	Error on starting server: [Exception... "Component returned failure code: 0x804b0036 (NS_ERROR_SOCKET_ADDRESS_IN_USE) [nsIServerSocket.initSpecialConnection]"  nsresult: "0x804b0036 (NS_ERROR_SOCKET_ADDRESS_IN_USE)"  location: "JS frame :: chrome://marionette/content/server.js :: MarionetteServer.prototype.start :: line 85"  data: no]
[Exception... "Component returned failure code: 0x804b0036 (NS_ERROR_SOCKET_ADDRESS_IN_USE) [nsIServerSocket.initSpecialConnection]"  nsresult: "0x804b0036 (NS_ERROR_SOCKET_ADDRESS_IN_USE)"  location: "JS frame :: chrome://marionette/content/server.js :: MarionetteServer.prototype.start :: line 85"  data: no]
MarionetteServer.prototype.start@chrome://marionette/content/server.js:85:19
MarionetteComponent.prototype.init@resource://gre/components/marionettecomponent.js:218:5
MarionetteComponent.prototype.observe@resource://gre/components/marionettecomponent.js:142:7
```

It seems clear trace logs it is better to check from geckodriver first. And selenium developers once checked, it needs to route to geckodriver and check why it always have such error which cause connection failed.

**Comment by davehunt** [comment]
> Specifically, your error shows that webdriver's `UnexpectedAlertPresentException` is thrown and not handled.

This is not a correct interpretation of the stack trace. The exception raised is _after_ the `elif` block checking the exception class against `UnexpectedAlertPresentException`.

@HackToday your geckodriver log suggests that it's failed to start because the address is already in use. See references to `NS_ERROR_SOCKET_ADDRESS_IN_USE`.

Looking at your log, I can see the geckodriver is started 200 times, each with a unique port. Marionette is started 193 times, each of these also with a unique port. There are no ports common between geckodriver and marionette. The last port selected by geckodriver is 52295, and the failure occurs when attempting to connect to marionette on port 59872.

It would seem that geckodriver has chosen a port for marionette to listen on, which is unavailable at the time marionette is started. Do you have anything running on this port? Perhaps there is a race condition when both geckodriver and marionette are selecting available ports?

**Comment by andreastt** [comment]
There is certainly a chance of a race condition happening the way Marionette is currently assigned a port: When Selenium/the consumer of geckodriver picks a port it’s usually done by binding to port 0 which atomically returns a free port from the system. It then releases this port and passes the port number onto geckodriver which instructs Marionette/Firefox to bind to it again. In the interim there’s a chance something else might have bound to that port, causing `NS_ERROR_SOCKET_ADDRESS_IN_USE`.

See https://bugzilla.mozilla.org/show_bug.cgi?id=1240830 for a discussion on various ideas for solving this race in the long term.

Since I don’t consider this a geckodriver specific bug, I’m going to go ahead and close this issue.

**Comment by HackToday** [comment]
@davehunt  Thanks for your reply.  
I did not have anything running against on 59872, as you can find from `netstat`, Only `geckodriver and Marionette` change port randomly from time to time.  

And also I tried one new experiment, right now, I choose Remote webdriver, instead of Firefox local webDriver, It seems Remote webdriver better than Firefox webdriver. (It not hit such `NS_ERROR_SOCKET_ADDRESS_IN_USE` issue anymore, At least, in several run test now. )

In summary, I should say Firefox and geckodriver still have some unexpected issue which can break tests from time to time.
